### PR TITLE
Get ACF version in reliable manner.

### PR DESCRIPTION
**Pull request by herregroen**
[ACF will never downgrade the version in the database.](https://github.com/elliotcondon/acf/blob/master/core/controllers/upgrade.php#L64-L69)

As such using get_option when you've tried out ACF Pro but later downgraded back to the free version will still return a version greater than 5.

Instead we should use either a function that only exists in ACF Pro or a filter that only exists in ACF 4 to get the real version.

### CODE REVIEW

File: inc/configuration/class-yoast-acf-analysis-configuration.php [hunk 1]
```diff
@@ -48,7 +48,11 @@ public function __construct(
 	 * @return string The ACF version.
 	 */
 	public function get_acf_version() {
-		return get_option( 'acf_version' );
+		if ( function_exists( 'acf_get_setting') ) {
```
Comment (kraftner) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Missing space:
`if ( function_exists( 'acf_get_setting' ) ) {`
`if ( function_exists( 'acf_get_setting') ) {`

File: inc/configuration/class-yoast-acf-analysis-configuration.php [hunk 2]
```diff
@@ -48,7 +48,11 @@ public function __construct(
...
+			return acf_get_setting( 'version' );
+		}
+
+		return apply_filters('acf/get_info', 'version');
```
Comment (kraftner) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Missing spaces:
`return apply_filters('acf/get_info', 'version');`
`return apply_filters( 'acf/get_info', 'version' );`

**Comment by kraftner** [comment]
Okay I've now finally had a time to look into this in more detail. Some more things to improve:

- [This test fails](https://github.com/Yoast/yoast-acf-analysis/blob/7e3df0c73f302393febe5bf88d31e14858239fbb/tests/php/unit/Configuration/ConfigurationTest.php#L16-L45) and needs to be updated
- There are two more places where the version is determined via the option that should be changed as well
https://github.com/Yoast/yoast-acf-analysis/blob/7e3df0c73f302393febe5bf88d31e14858239fbb/inc/class-ac-yoast-acf-content-analysis.php#L85
https://github.com/Yoast/yoast-acf-analysis/blob/7e3df0c73f302393febe5bf88d31e14858239fbb/inc/class-ac-yoast-acf-content-analysis.php#L210
- After those two changes [this test will also fail](https://github.com/Yoast/yoast-acf-analysis/blob/7e3df0c73f302393febe5bf88d31e14858239fbb/tests/php/unit/MainTest.php#L15-L33)

**Comment by kraftner** [comment]
Thinking some more about this I think that getting the version at three different places isn't very DRY and prone to exactly what happened in this PR: Changing behaviour at only one spot resulting in inconsistencies.

File: inc/configuration/class-yoast-acf-analysis-configuration.php [hunk 3]
```diff
@@ -48,7 +48,11 @@ public function __construct(
 	 * @return string The ACF version.
 	 */
 	public function get_acf_version() {
-		return get_option( 'acf_version' );
+		if ( function_exists( 'acf_get_setting' ) ) {
```
Comment (moorscode) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Could you document in what case this is applicable and why it is needed?

File: tests/php/unit/Configuration/ConfigurationTest.php [hunk 4]
```diff
@@ -42,6 +34,7 @@ public function testEmpty() {
 			$configuration->to_array()
 		);
 
+		$this->assertEquals( Filters\applied( 'acf/get_info' ), 1 );
```
Comment (kraftner) [inline tests/php/unit/Configuration/ConfigurationTest.php]: This only tests the code path for ACF4. Should probably also test the case inside `\Yoast_ACF_Analysis_Configuration->get_acf_version()` that `acf_get_setting()` exists.

File: inc/class-ac-yoast-seo-acf-content-analysis.php [hunk 5]
```diff
@@ -80,7 +80,11 @@ public function boot() {
 	 * Boots the plugin for dev environment.
 	 */
 	public function boot_dev() {
-		$version = ( -1 === version_compare( get_option( 'acf_version' ), 5 ) ) ? '4' : '5';
+		$registry = Yoast_ACF_Analysis_Facade::get_registry();
+		$configuration = $registry->get( 'config' );
+
+		$version = ( -1 === version_compare( $configuration->get_acf_version(), 5 ) ) ? '4' : '5';
```
Comment (jcomack) [inline inc/class-ac-yoast-seo-acf-content-analysis.php]: This shorthand if-statement doesn't seem necessary (and readable).

I'd prefer defaulting to version 4, and only checking if we need to bump to version 5 via `version_compare`.

**Comment by jcomack** [comment]
CR done 👍

**Comment by kraftner** [comment]
Just for completeness this still leaves the theoretical risk that something about the data format [gets upgraded ACF4->5](https://plugins.trac.wordpress.org/browser/advanced-custom-fields/tags/5.6.5/includes/admin/install-updates.php) that later can't be read properly any more when going back to ACF4 as ACF doesn't do any backwards migrations. Just leaving this here in case this might pop up later.

But generally this fixes the issue and is a clear improvement over the current state. This is why I am merging this. :tada:

**Pull request merged**